### DISCUSSION TITLE
Is it planned to support both rv32imac and bare rv32i without these extensions?

### DISCUSSION THREAD

**Discussion by Plaque-fcc** [micropython/micropython#15964] [Core Development]
Hi!

  I'm quite new to the development of such things, but after me constantly mentioning MicroPython here and there on different devices, my friend said that he wants to give it a try.
  He's very good at Python, but all that he currently has is a j2me-enabled phone, a rv32i emulator in Java for j2me, and a QEMU setup. Building it is on my part.

  Could someone please explain what to isolate and where to separate to make mpy-cross (presumably we need changes there) support rv32i with or without various extensions? Appreciate any help.

**Comment by Plaque-fcc** [comment]
Can actually have it with these additions:
https://github.com/Plaque-fcc/micropython/tree/qemu-rv32i

Standalone commits:
[01_commit-3fcf4e8.txt](https://github.com/user-attachments/files/17264003/01_commit-3fcf4e8.txt)
[00_commit-88ae530.txt](https://github.com/user-attachments/files/17264005/00_commit-88ae530.txt)

**Reply by Plaque-fcc** [reply]
More from my friend: https://github.com/orgs/micropython/discussions/15980

**Reply by agatti** [reply]
Both extensions are required for the native/viper code emitters to work.  If you don't need that then you can probably change the makefiles to use plain rv32i and explicitly disable `MICROPY_EMIT_RV32`.

For the `M` extension, one of the operations needed to generate code for is in fact multiplication.  You can replace the implementation of `ASM_MUL_REG_REG` in `py/asmrv32.h` with an emitter for a series of shifts and additions.  Unless there's a hardware RV32-based platform MicroPython runs on that doesn't support the `M` extension, I'd say that such a thing won't probably show up in the codebase.

The `C` extension requirement has a similar background: the first iteration of the native code generator only supported `I` and `M` extensions, but after some discussion then the `C` extension was made a requirement.  The only real hardware platform MicroPython ran on that supported RV32 at the time (ESP32C3) has an IMC-enabled core, so it made sense to use that.

Making the `C` extension optional isn't that difficult, it's just a matter of placing the right ifdefs in `py/asmrv32.c`.  However the same remark about seeing such a change merged in still applies.  Technically the same can be said about the `I` extension too, as some RV32E MCUs do exist, but none of them run MicroPython (as far as I know).
